### PR TITLE
[Intel MKL] Implement new DNNL1.x MatMul primitive cache

### PR DESCRIPTION
**Pull request by Zantares**
This PR has done:
* Implement primive cache for new primitive `MatMul` in DNNL 1.x. 
* Remove redundant `Reorder` and change destination type for BF16 MatMul, it helps to enable primitive cache for BF16 MatMul. 

`MatMul` has similar semantic with InnerProduct, but they have different parameters and attributes in DNNL. Current we just use it for non-fused MatMul in TF.

Signed-off-by: Lu Teng teng.lu@intel.com

### CODE REVIEW

File: tensorflow/core/kernels/mkl_matmul_ops_common.h [hunk 1]
```diff
@@ -516,33 +516,191 @@ class MklDnnMatMulOpBase : public OpKernel {
...
+    // Create memory primitive based on dummy data.
+    context_.a_mem.reset(
+        new mkldnn::memory(*context_.a_md, cpu_engine_, DummyData));
+    context_.b_mem.reset(
+        new mkldnn::memory(*context_.b_md, cpu_engine_, DummyData));
+    context_.c_mem.reset(
+        new mkldnn::memory(*context_.b_md, cpu_engine_, DummyData));
+
+    // Create matmul primitive.
+    context_.matmul_prim.reset(new mkldnn::matmul(*context_.prim_desc));
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_ops_common.h:633]: I think `matmul_prim` can just be a local variable here (doesn't need to be a class member). Only `matmul_primitives` is used in `Execute`.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and I'm so sorry for the delay! I have minor comments.

**Comment by gbaned** [comment]
@Zantares Can you please check @penpornk's comments and keep us posted. Thanks!

File: tensorflow/core/kernels/mkl_matmul_ops_common.h [hunk 2]
```diff
@@ -516,33 +516,191 @@ class MklDnnMatMulOpBase : public OpKernel {
...
+  }
+
+  ~MklMatMulPrimitive() {}
+
+  void Execute(const T* a_data, const T* b_data, T* c_data) {
+    context_.a_mem->set_data_handle(static_cast<void*>(const_cast<T*>(a_data)));
+    context_.b_mem->set_data_handle(static_cast<void*>(const_cast<T*>(b_data)));
+    context_.c_mem->set_data_handle(static_cast<void*>(const_cast<T*>(c_data)));
+
+    execute_primitives(context_.matmul_primtimives, context_.stream,
```
[the remark this answers is not in the document]
Comment (Zantares) [reply tensorflow/core/kernels/mkl_matmul_ops_common.h]: Done.

File: tensorflow/core/kernels/mkl_matmul_ops_common.h (hunk 1 above)
Comment (Zantares) [reply tensorflow/core/kernels/mkl_matmul_ops_common.h]: Done.

**Comment by Zantares** [comment]
made a rebase and addressed the code, please take a look, thanks!

**Review by penpornk — APPROVED** [review]
Thank you for the changes!

**Pull request merged**